### PR TITLE
[Merged by Bors] - Introduce a queue for attestations from the network

### PR DESCRIPTION
**Pull request by paulhauner**
## Issue Addressed

N/A

## Proposed Changes

Introduces the `GossipProcessor`, a multi-threaded (multi-tasked?), non-blocking processor for some messages from the network which require verification and import into the `BeaconChain`.

Initial testing indicates that this massively improves system stability by (a) moving block tasks from the normal executor (b) spreading out attestation load.

## Additional Info

TBC

### CODE REVIEW

File: beacon_node/network/src/router/gossip_processor.rs [hunk 1]
```diff
@@ -0,0 +1,810 @@
...
+                metrics::set_gauge(
+                    &metrics::GOSSIP_PROCESSOR_UNAGGREGATED_ATTESTATION_QUEUE_TOTAL,
+                    attestation_queue.len() as i64,
+                );
+                metrics::set_gauge(
+                    &metrics::GOSSIP_PROCESSOR_AGGREGATED_ATTESTATION_QUEUE_TOTAL,
+                    aggregate_queue.len() as i64,
+                );
+
+                if initial_aggregate_queue_len != aggregate_queue.len() && aggregate_queue.is_full()
```
Comment (michaelsproul) [inline beacon_node/network/src/router/gossip_processor.rs:334]: I don't understand the first half of this condition... is this so that we only print the message when the queue fills up for the first time, and not every subsequent time. I'd almost prefer a `aggregate_queue.len() > initial_aggregate_queue_len` to make the intention clearer

File: beacon_node/network/src/router/gossip_processor.rs [hunk 2]
```diff
@@ -0,0 +1,810 @@
...
+                            let _attestation_timer = metrics::start_timer(
+                                &metrics::GOSSIP_PROCESSOR_UNAGGREGATED_ATTESTATION_WORKER_TIME,
+                            );
+                            metrics::inc_counter(
+                                &metrics::GOSSIP_PROCESSOR_UNAGGREGATED_ATTESTATION_VERIFIED_TOTAL,
+                            );
+
+                            let beacon_block_root = attestation.data.beacon_block_root;
+
+                            let attestation = if let Ok(attestation) = chain
```
Comment (michaelsproul) [inline beacon_node/network/src/router/gossip_processor.rs:406]: This `if let Ok(_)` makes it look like we're dropping the error message, might be clearer to write it as a match that plumbs the error into `handle_attestation_verification_failure`?

Comment (michaelsproul) [reply beacon_node/network/src/router/gossip_processor.rs:406]: I won't block on it though

File: beacon_node/network/src/router/gossip_processor.rs [hunk 3]
```diff
@@ -0,0 +1,810 @@
...
+
+                            if !should_import {
+                                return;
+                            }
+
+                            metrics::inc_counter(
+                                &metrics::GOSSIP_PROCESSOR_UNAGGREGATED_ATTESTATION_IMPORTED_TOTAL,
+                            );
+
+                            if let Err(e) = chain.apply_attestation_to_fork_choice(&attestation) {
```
Comment (michaelsproul) [inline beacon_node/network/src/router/gossip_processor.rs:435]: What about the naive aggregation pool? Don't we also need to add it to that?

File: beacon_node/network/src/router/gossip_processor.rs [hunk 4]
```diff
@@ -0,0 +1,810 @@
...
+                                &metrics::GOSSIP_PROCESSOR_AGGREGATED_ATTESTATION_WORKER_TIME,
+                            );
+                            metrics::inc_counter(
+                                &metrics::GOSSIP_PROCESSOR_AGGREGATED_ATTESTATION_VERIFIED_TOTAL,
+                            );
+
+                            let beacon_block_root =
+                                boxed_aggregate.message.aggregate.data.beacon_block_root;
+
+                            let aggregate = if let Ok(aggregate) = chain
```
Comment (michaelsproul) [inline beacon_node/network/src/router/gossip_processor.rs:470]: As above with the `if let`

**Comment by paulhauner** [comment]
> I also have a general question about the use of a LIFO queue rather than a FIFO queue.

I agree that the choice of queue type is not clear-cut, but I can share my thinking.

There are three reasons we want attestations:

### 1. Update our view of the fork choice.

If we assume no forks and no slashings, an attestation at epoch `n + 1` is strictly more valuable than one at epoch `n`, since the vote on a higher block in the chain covers all previous blocks (including the one voted on in `n`). This makes me think that LIFO is best here, even when we have forks and slashings.

### 2. Learn about new blocks that we didn't already get on gossip.

Naively I would say LIFO since we learn about high blocks earlier, which will eventually lead us to all ancestors of that block.

That being said, I'm sure about how this would play-out in the real world since you need to download a whole block before you learn about the parent.

### 3. Store for inclusion in blocks.

It's not immediately clear to me what type of queue is best here... You don't want old attestations since you wont be the first to include and wont get the reward, but due to `MIN_ATTESTATION_INCLUSION_DELAY` you don't want them to be *too* new. Perhaps the FIFO with front-drop suits here?

So, in summary, I'm really not sure what's going to work best here. FIFO with front-pop is a good idea, I'm torn between that and LIFO. Given this information, do you have any preference?

File: beacon_node/network/src/router/gossip_processor.rs (hunk 3 above)
Comment (paulhauner) [reply beacon_node/network/src/router/gossip_processor.rs:435]: Great catch!

File: beacon_node/network/src/router/gossip_processor.rs [hunk 5]
```diff
@@ -0,0 +1,810 @@
...
+        // The manager future will run on the non-blocking executor and delegate tasks to worker
+        // threads on the blocking executor.
+        let manager_future = async move {
+            loop {
+                // Listen to both the event and idle channels, acting on whichever is ready
+                // first.
+                //
+                // Set `work_event = Some(event)` if there is new work to be done. Otherwise sets
+                // `event = None` if it was a worker becoming idle.
+                let work_event;
```
[the remark this answers is not in the document]
Comment (paulhauner) [reply beacon_node/network/src/router/gossip_processor.rs:207]: Turns out it does.. I was fighting it earlier but it must have been unrelated. Pretty dank macro that one.

File: beacon_node/network/src/router/gossip_processor.rs (hunk 1 above)
Comment (paulhauner) [reply beacon_node/network/src/router/gossip_processor.rs]: Oh yeah, this was a bit of a lazy effort to get de-bounce on logs. I've added something better here: b95fa5b.

**Comment by paulhauner** [comment]
All comments addressed, let me know your thoughts on queue type :)

File: beacon_node/network/src/router/gossip_processor.rs (hunk 1 above)
Comment (michaelsproul) [reply beacon_node/network/src/router/gossip_processor.rs]: I almost preferred the simple condition haha, but I guess we could factor the fancy debouncer out and use it elsewhere at some point

**Comment by paulhauner** [comment]
Yay! Thank you!

bors r+
